director/cache_ads_test.go

```diff
@@ -176,3 +180,62 @@ func TestGetAdsForPath(t *testing.T) {
 	assert.Equal(t, len(oAds), 0)
 	assert.Equal(t, len(cAds), 0)
 }
+
+func TestConfigCacheEviction(t *testing.T) {
+	mockPelicanOriginServerAd := ServerAd{
+		Name:    "test-origin-server",
+		AuthURL: url.URL{},
+		URL: url.URL{
+			Scheme: "https",
+			Host:   "fake-origin.org:8443",
+		},
+		WebURL: url.URL{
+			Scheme: "https",
+			Host:   "fake-origin.org:8444",
+		},
+		Type:      OriginType,
+		Latitude:  123.05,
+		Longitude: 456.78,
+	}
+	mockNamespaceAd := NamespaceAd{
+		RequireToken:  true,
+		Path:          "/foo/bar/",
+		Issuer:        url.URL{},
+		MaxScopeDepth: 1,
+		Strategy:      "",
+		BasePath:      "",
+		VaultServer:   "",
+	}
+
+	t.Run("evicted-origin-can-cancel-health-test", func(t *testing.T) {
+		serverAds.DeleteAll()
+		// Clear the map for the new test
+		healthTestCancelFuncs = make(map[ServerAd]context.CancelFunc)
+		ConfigCacheEviction()
+
+		serverAds.Set(mockPelicanOriginServerAd, []NamespaceAd{mockNamespaceAd}, ttlcache.DefaultTTL)
+		ctx, cancelFunc := context.WithDeadline(context.Background(), time.Now().Add(time.Second*5))
+		healthTestCancelFuncs[mockPelicanOriginServerAd] = cancelFunc
+
+		require.True(t, serverAds.Has(mockPelicanOriginServerAd), "serverAds failed to register the originAd")
+
+		cancelReceived := false
+		go func() {
+			<-ctx.Done()
+			if ctx.Err() == context.Canceled {
+				cancelReceived = true
+			}
+		}()
+
+		serverAds.Delete(mockPelicanOriginServerAd) // This should call onEviction handler and close the context
+
+		require.False(t, serverAds.Has(mockPelicanOriginServerAd), "serverAds didn't delete originAd")
+
+		// OnEviction is handled on a different goroutine than the cache management
+		// So we want to wait for a bit so that OnEviction can have time to be
+		// executed
+		time.Sleep(2 * time.Second) // This is a bit flaky and can delay the test run, so any thoughts on improvement is very welcome
```

Comment:
As comment says, any thoughts on improving this is very welcomed.

Reply:
I would suggest something like this:

```
                 cancelChan := make(chan int)
		go func() {
			<-ctx.Done()
			if ctx.Err() == context.Canceled {
				cancelChan <- 1
			}
		}()

                 ...

                 select {
                 case <-cancelChan:
                 	require.True(t, true)
                 case <-time.After(3 * time.Second):
                 	require.True(t, false)
                 }
```

So, instead of polling (and in this case, once), you get a notification via the channel from the other go routine that the eviction is done.  It also includes a reasonable timeout so, on failure, the test doesn't hang forever.